packages/grid/x-grid/src/tests/selection.XGrid.test.tsx

```diff
@@ -126,4 +133,30 @@ describe('<XGrid /> - Selection', () => {
     });
     expect(apiRef.current.getSelectedRows()).to.have.keys([0]);
   });
+
+  it('should select only filtered rows after filter is applied', () => {
+    render(<Test checkboxSelection />);
+    const selectAll = screen.getByRole('checkbox', {
+      name: /select all rows checkbox/i,
+    });
+    apiRef.current.setFilterModel({
+      items: [
+        {
+          columnField: 'brand',
+          operatorValue: 'contains',
+          value: 'Puma',
+        },
+      ],
+    });
+    expect(getColumnValues(1)).to.deep.equal(['Puma']);
+    fireEvent.click(selectAll);
+    // TODO fix, should be only 2
+    expect(Array.from(apiRef.current.getSelectedRows().keys())).to.deep.equal([0, 1, 2]);
```

Comment:
fine to use apiRef to assert the state but your test coverage would be higher if you check the `Mui-selected` css class on the row

Reply:
My bad, no, I got trapped. We can't, the selected rows 0 and 1 are not visible.